Add support for local TTS

FoxCunning opened the pull request:
This change implements support for local/system text-to-speech using the [Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API).

It will read the generated text up to 500 characters. If the prompt has been modified, it will also read the portion that was changed.

![image](https://github.com/user-attachments/assets/34bc798a-fb15-42f2-b188-56f4ef886832)

lmg-anon commented:
This is interesting, but I don't think it should be a new collapsible group in the sidebar. Maybe it could open a new modal when clicking on a button here instead:  
![image](https://github.com/user-attachments/assets/664e9140-9eab-47a2-9e7b-91b1a08f4aaa)

Or, even better, the configurations could be added to the `Editor Preferences` modal.

FoxCunning commented:
The Editor Preferences modal sounds good and I'll see if I can move all the settings there.
There wouldn't be a button to stop the active playback on the main UI, though.

lmg-anon commented on mikupad.html:
```diff
@@ -1686,9 +1716,12 @@
 	});
 	if (!res.ok)
 		throw new Error(`HTTP ${res.status}`);
+	window.TTS.new_prompt = options.prompt;
+	window.TTS.new_text = "";
 	for await (const chunk of parseEventStream(res.body)) {
 		const probs = chunk.completion_probabilities[0]?.probs ?? [];
 		const prob = probs.find(p => p.tok_str === chunk.content)?.prob;
+		window.TTS.new_text += chunk.content;
```
Please don't add code like this to `llamaCppCompletion`; it ties this functionality to the llama.cpp backend. Instead, add the code to the `predict` method in the `App` component. You could also add a `useEffect` with `[promptText, cancel]` as dependencies, so it gets executed after a prediction is finished.

FoxCunning replied:
Usin React hooks is a great idea. I'll try that, thanks.

FoxCunning replied:
Well, it was a good idea, but unfortunately it seems to be very unreliable. Both with promptText and promptChunks, it randomly misses a few tokens, most often the last one.
I'll have to add a few lines to App.predict instead. I'll try to modify it as little as possible, and instead add tts-specific methods to be called from there.

lmg-anon commented on mikupad.html:
```diff
@@ -1697,6 +1730,29 @@
 			} : {})
 		};
 	}
+	// Check if something has been added to the prompt, and read that too
+	if (window.TTS.prev_prompt != window.TTS.new_prompt) {
+		var prev = window.TTS.prev_prompt;
+		var next = window.TTS.new_prompt;
+		if (window.TTS.prev_prompt.length > 2000) { prev = window.TTS.prev_prompt.substr(1500) }
+		if (window.TTS.new_prompt.length > 2000) { next = window.TTS.new_prompt.substr(1500) }
+		window.TTS.prev_prompt = options.prompt + window.TTS.new_text;
+		// Find where the added/modified part begins
+		for (var c = 0; c < Math.min(prev.length, next.length); c++) {
+			if (prev[c] != next[c]) {
+				break;
+			}
+		}
+		if (c < next.length) {
+			window.TTS.new_text = next.substr(c) + window.TTS.new_text;				
+		}
+	} else {
+		window.TTS.prev_prompt = options.prompt + window.TTS.new_text;
+	}
+	// Trim if string is too long
+	if (window.TTS.new_text.length > 500) { window.TTS.new_text = window.TTS.new_text.substr(-500) }
+	// Read out TTS.new_text
+	if (window.TTS.enabled) { setTimeout(textToSpeech, 20) }
```
Same as above.

lmg-anon commented on mikupad.html:
```diff
@@ -5218,6 +5274,11 @@
 	const [sessionEndpointConnecting, setSessionEndpointConnecting] = useState(false);
 	const [sessionEndpointError, setSessionEndpointError] = useState(undefined);
 	const [showAPIKey, setShowAPIKey] = useState(false);
+	const [ttsEnabled, setTTSEnabled] = usePersistentState('ttsEnabled', 0);
+	const [ttsVoice, setTTSVoice] = usePersistentState('ttsVoice', 0);
+	const [ttsPitch, setTTSPitch] = usePersistentState('ttsPitch', 1);
+	const [ttsRate, setTTSRate] = usePersistentState('ttsRate', 1);
+	const [ttsVolume, setTTSVolume] = usePersistentState('ttsVolume', 1);
```
These states aren't being applied to window.TTS

lmg-anon commented on mikupad.html:
```diff
@@ -7130,6 +7229,26 @@
 					Export prompt to plaintext
 				</button>
 			</div>
+			<${Checkbox} label="Enable Text-to-Speech"
+						disabled=${!!cancel} value=${window.TTS.enabled} onValueChange=${changeTTSEnabled}/>
+			<div className="hbox-flex" style=${{ "flex-wrap": "unset" }}>
+					<${SelectBox}
+						id="voices"
+						label="Voice"
+						disabled=${!!cancel}
+						value=${window.TTS.voice_id}
+						onValueChange=${selectNewVoice}
+						options=${listTTSVoices}/>
+					<button title="Stop TTS" className="symbol-button" disabled=${cancel} onClick=${ttsStop}>
+						<${SVG_Cancel} style=${{ 'width': '.95em', 'transform': 'translate(-50%, -45%)' }}/>
+					</button>
+				</div>
+			<${InputSlider} label="TTS Pitch" type="number" step="0.1" max="2"
+				readOnly=${!!cancel} value=${window.TTS.pitch} onValueChange=${changeTTSPitch}/>
+				<${InputSlider} label="TTS Rate" type="number" step="0.1" max="2"
+				readOnly=${!!cancel} value=${window.TTS.rate} onValueChange=${changeTTSRate}/>
+				<${InputSlider} label="TTS Volume" type="number" step="0.1" max="2"
```
This doesn't look good in the interface... I think a good improvement would be to hide the TTS-related elements when TTS is disabled.

FoxCunning commented:
Happy new year! 🎆🎉

I've now fully re-written the TTS code.

- The main change is in the App.predict method. Using "useEffect" hooks proved to be unreliable. On the bright side, I only had to add a few lines in that method.
- Almost all of the TTS functionality is in separate methods that I kept together starting at line 7350.
- The UI is in the Editor Preferences modal. Disabling TTS via checkbox makes all the other TTS-related elements disappear.
- I've also added an option to not narrate user inputs.
- window.TTS has been removed. Now it only uses React variables stored with useRef / usePersistentState inside the App class. TTS settings should be saved in local storage.
- The SVG I added is a simple "stop" button used to, well, stop the TTS while it's speaking. Since it's in the editor preferences, I also added a "hotkey" (CTRL+E) so it's possible to stop the narration without having to open the menu.

To be noted: the reason why I chose to process text chunks as soon as they come is that this way the speech synthesis can start narrating as soon as a sentence is complete (e.g. the AI generates a newline or other "stopping" token). This way, if a lot of text is being generated, and especially if the AI is slow, the user does not have to wait until it's all finished before the narration starts.
Unterminated user inputs will be narrated (if the option is selected) together with the next generation, to form a complete sentence.

I've tested it quite a bit with llama.cpp and koboldcpp.
If there's anything else you think should be changed, let me know.

FoxCunning reviewed (commented):
See previous comment in pull request.
